cscs-checks/microbenchmarks/strided_bandwidth/src/strides.cpp

```diff
@@ -0,0 +1,187 @@
+
+    std::chrono::duration<double> duration = t1 - t0;
+    return duration.count();
+}
+
+class Buffer
+{
+public:
+    Buffer() : data_size(0) {}
+
+    explicit Buffer(size_t sz_) : data_size(sz_)
+    {
+        data_size = (sz_/sizeof(size_t)) * (sizeof(size_t));
+
+        void* buf;
+        if (posix_memalign(&buf, 32, data_size)) {
+            std::cout << "alloc failed\n"; exit(1);
+        } 
+        data_.reset((size_t*)buf);
+
+        init();
+    }
+
+    Buffer(Buffer&& rhs) : data_(std::move(rhs.data_)), data_size(rhs.data_size) {}
+
+    Buffer& operator=(Buffer&& rhs) {
+        data_ = std::move(rhs.data_);
+        data_size = rhs.data_size;
+        return *this;
+    }
+
+    size_t* get() const { return data_.get(); }
+
+private:
+
+    void init()
+    {
+        for (size_t i = 0; i < data_size/sizeof(size_t); i++)
+            ((size_t*)data_.get())[i] = 1;
+    }
+
+    std::unique_ptr<size_t> data_;
+    size_t data_size;
+};
+
+
+void set_affinity(std::thread& t, int i)
+{
+    cpu_set_t cpuset;
+    CPU_ZERO(&cpuset);
+    CPU_SET(i, &cpuset);
+    int rc = pthread_setaffinity_np(t.native_handle(),
+                                    sizeof(cpu_set_t), &cpuset);
+    if (rc != 0)
+        std::cerr << "Error calling pthread_setaffinity_np: " << rc << "\n";
+}
+
+int get_affinity(std::thread& t)
+{
+    cpu_set_t cpuset;
+    CPU_ZERO(&cpuset);
+    int rc = pthread_getaffinity_np(t.native_handle(),
+                                    sizeof(cpu_set_t), &cpuset);
+    if (rc != 0)
+        std::cerr << "Error calling pthread_setaffinity_np: " << rc << "\n";
+    int ret;
+    for (int i=0; i < 128; ++i)
+        if (CPU_ISSET(i, &cpuset)) return i;
+
+    return -1;
+}
+
+void update_stride(size_t*, size_t, size_t, size_t);
+
+
+int main(int argc, char ** argv)
+{
+    if (argc != 4)
+    {
+        std::cout << "Usage: ./<prog_name> buffer_size stride nthreads\n"
+                     "\n"
+                     "buffer_size: in bytes\n"
+                     "stride: in multiples of 8 bytes\n";
+        exit(1);
+    }
+
+    size_t s = std::stoi(argv[1]);
+    size_t stride = std::stoi(argv[2]);
+    int nthreads = std::stoi(argv[3]);
+
+    double time_per_run = 2.0;
+
+    std::vector<Buffer> buf(nthreads);
+    // initialize buffers (first touch)
+    for (int i = 0; i < nthreads; ++i)
+    {
+        std::thread t([=, &buf](){
+                std::this_thread::sleep_for(std::chrono::milliseconds(20));
+                buf[i] = Buffer(s);
+            });
+        set_affinity(t, i);
+        t.join();
+    }
+
+    std::function<void(size_t)> func = std::bind(&update_stride, buf[0].get(), s, std::placeholders::_1, stride);
+    size_t loops = determine_loops(func, time_per_run);
+
+    std::vector<std::function<void()>> tasks(nthreads);
+    for (int i = 0; i < nthreads; ++i)
+        tasks[i] = std::bind(&update_stride, buf[i].get(), s, loops, stride);
```

Comment:
Similarly here, I'd rather use a lambda.

Reply:
The lambda would be cumbersome here, since I want to put the resulting objects in a container.
I guess decltype could be used, but I don't think it's worth the hassle. The lambda
is also not as readable here, since I bind all arguments and don't leave any placeholders.
Performance-wise, the additional function call of std::function is irrelevant since it's called a single time.